File: src/modules/digital-twin/server/schemas/insar.py
from datetime import date, datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field


class InsarJobCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    date_from: date
    date_to: date
    reference_date: date | None = None
    orbit_direction: str | None = Field(None, pattern="^(ASC|DESC|asc|desc)$")
    aoi_geojson: dict[str, Any] | None = Field(
        None, description="GeoJSON Polygon; default = boundary do projeto"
    )
    run_immediately: bool = True


class InsarJobOut(BaseModel):
    id: UUID
    project_id: UUID
    name: str
    status: str
    date_from: date
    date_to: date
    reference_date: date | None
    orbit_direction: str | None
    scene_count: int
    error_message: str | None
    properties: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime
    updated_at: datetime
    completed_at: datetime | None = None
    aoi: dict[str, Any] | None = None


class InsarJobListResponse(BaseModel):
    items: list[InsarJobOut]
    total: int


class InsarRasterOut(BaseModel):
    id: UUID
    project_id: UUID
    job_id: UUID
    insar_image_id: UUID | None
    raster_kind: str
    epoch_date: date | None
    file_path: str
    download_url: str
    preview_url: str | None = None
    metadata_url: str
    crs_epsg: int
    pixel_size_m: float | None
    width: int | None
    height: int | None
    units: str
    min_value: float | None
    max_value: float | None
    mean_value: float | None
    footprint: dict[str, Any] | None
    properties: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime


class InsarRasterListResponse(BaseModel):
    items: list[InsarRasterOut]
    total: int


class InsarRasterMetadata(BaseModel):
    id: UUID
    raster_kind: str
    epoch_date: date | None
    crs_epsg: int
    pixel_size_m: float | None
    width: int | None
    height: int | None
    units: str
    nodata_value: float | None
    min_value: float | None
    max_value: float | None
    mean_value: float | None
    footprint: dict[str, Any] | None
    download_url: str
    properties: dict[str, Any] = Field(default_factory=dict)

File: src/modules/digital-twin/viewer/api/insar.ts
import { resolveGeoApiBase } from "./geo-base";
import { apiUrl } from "@/lib/api-url";

export interface InsarJobCreate {
  name: string;
  date_from: string;
  date_to: string;
  reference_date?: string;
  orbit_direction?: "ASC" | "DESC";
  aoi_geojson?: GeoJSON.Polygon;
  run_immediately?: boolean;
}

export interface InsarJob {
  id: string;
  project_id: string;
  name: string;
  status: string;
  date_from: string;
  date_to: string;
  scene_count: number;
  properties: Record<string, unknown>;
}

export interface InsarRaster {
  id: string;
  raster_kind: string;
  epoch_date?: string | null;
  download_url: string;
  preview_url?: string | null;
  metadata_url: string;
  min_value?: number | null;
  max_value?: number | null;
  units: string;
}

function geoInsarUrl(projectId: string, suffix: string): string {
  const base = resolveGeoApiBase();
  const pid = encodeURIComponent(projectId);
  return `${base}/v1/projects/${pid}/insar${suffix}`;
}

export async function createInsarJob(
  projectId: string,
  body: InsarJobCreate,
): Promise<InsarJob> {
  const res = await fetch(geoInsarUrl(projectId, "/jobs"), {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  if (!res.ok) throw new Error(await res.text());
  return res.json() as Promise<InsarJob>;
}

export interface InsarRasterDetail extends InsarRaster {
  footprint?: GeoJSON.Polygon | null;
  crs_epsg?: number;
  nodata_value?: number | null;
}

export async function listInsarRasters(
  projectId: string,
  jobId?: string,
  kind?: "displacement" | "velocity" | "coherence",
): Promise<{ items: InsarRaster[] }> {
  const params = new URLSearchParams();
  if (jobId) params.set("job_id", jobId);
  if (kind) params.set("kind", kind);
  const qs = params.toString();
  const res = await fetch(
    geoInsarUrl(projectId, `/rasters${qs ? `?${qs}` : ""}`),
  );
  if (!res.ok) throw new Error(await res.text());
  return res.json() as Promise<{ items: InsarRaster[] }>;
}

export async function fetchInsarRasterDetail(
  projectId: string,
  rasterId: string,
): Promise<InsarRasterDetail> {
  const res = await fetch(
    geoInsarUrl(projectId, `/rasters/${encodeURIComponent(rasterId)}`),
  );
  if (!res.ok) throw new Error(await res.text());
  return res.json() as Promise<InsarRasterDetail>;
}

export async function retryInsarJob(
  projectId: string,
  jobId: string,
): Promise<{ ok: boolean; jobId: string }> {
  const res = await fetch(
    apiUrl(
      `/api/geo/v1/projects/${encodeURIComponent(projectId)}/insar/jobs/${encodeURIComponent(jobId)}/run`,
    ),
    { method: "POST" },
  );
  if (!res.ok) throw new Error(await res.text());
  return res.json() as Promise<{ ok: boolean; jobId: string }>;
}

export function insarGeotiffUrl(projectId: string, rasterId: string): string {
  const path = geoInsarUrl(
    projectId,
    `/rasters/${encodeURIComponent(rasterId)}/download`,
  );
  return apiUrl(path);
}

export type InsarPipelineStageLog = {
  step: string;
  at: string;
  ok: boolean;
  detail?: string;
};

export type InsarJobPollResponse = {
  id?: string;
  status: string;
  properties?: {
    stages?: InsarPipelineStageLog[];
    message?: string;
  };
  stages?: InsarPipelineStageLog[];
  error_message?: string | null;
};

export async function fetchInsarJobStatus(
  projectId: string,
  jobId: string,
): Promise<InsarJobPollResponse> {
  const res = await fetch(
    geoInsarUrl(projectId, `/jobs/${encodeURIComponent(jobId)}`),
  );
  if (!res.ok) throw new Error(await res.text());
  return res.json() as Promise<InsarJobPollResponse>;
}

export type InsarPreflightResponse = {
  obraId: number;
  obraNome: string;
  period: { date_from: string; date_to: string; orbit: string };
  aoi: { obra_geojson_aoi: boolean; obra_coordinates: boolean };
  catalog: {
    total_entries: number;
    ready_slc_local: number;
    pair_ready: boolean;
  };
  copernicus: {
    credentialsConfigured: boolean;
    usernameHint: string | null;
    tokenOk: boolean;
    expiresInSec?: number;
    tokenError?: string;
  };
  processing: {
    snap_configured: boolean;
    synthetic_fallback_enabled: boolean;
  };
  cesium: { note: string };
};

export async function fetchInsarPreflight(
  projectId: string,
  params: { date_from: string; date_to: string; orbit?: string },
): Promise<InsarPreflightResponse> {
  const q = new URLSearchParams({
    date_from: params.date_from,
    date_to: params.date_to,
  });
  if (params.orbit) q.set("orbit", params.orbit);
  const res = await fetch(
    geoInsarUrl(projectId, `/preflight?${q.toString()}`),
  );
  if (!res.ok) throw new Error(await res.text());
  return res.json() as Promise<InsarPreflightResponse>;
}
